File: src/main/java/xyz/emirdev/echogen/commands/SudoCommand.java
package xyz.emirdev.echogen.commands;

import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.tree.LiteralCommandNode;
import io.papermc.paper.command.brigadier.CommandSourceStack;
import io.papermc.paper.command.brigadier.Commands;
import io.papermc.paper.command.brigadier.argument.ArgumentTypes;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import xyz.emirdev.echogen.PluginCommand;

import java.util.List;
import java.util.Set;

public class SudoCommand extends PluginCommand {

    @Override
    public LiteralCommandNode<CommandSourceStack> getCommand() {
        return Commands.literal("sudo")
                .requires(hasPermission("echogen.sud"))
                .then(Commands.argument("players", ArgumentTypes.players())
                        .then(Commands.argument("message", StringArgumentType.greedyString())
                                .executes(this::sudo)))
                .build();
    }

    private int sudo(CommandContext<CommandSourceStack> ctx) throws CommandSyntaxException {
        CommandSender sender = ctx.getSource().getSender();
        List<Player> targets = getPlayers(ctx, "players");
        if (targets == null)
            return 1;

        String message = StringArgumentType.getString(ctx, "message");

        for (Player target : targets) {
            target.chat(message);

            sender.sendRichMessage(
                    "<aqua>You have made <dark_aqua><player></dark_aqua> send <dark_aqua><message></dark_aqua></aqua>",
                    Placeholder.unparsed("player", target.getName()),
                    Placeholder.unparsed("message", message));
        }

        return 1;
    }
}

File: src/main/kotlin/cat/emir/echogen/commands/SudoCommand.kt
package cat.emir.echogen.commands

import cat.emir.echogen.Echogen
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.context.CommandContext
import io.papermc.paper.command.brigadier.CommandSourceStack
import io.papermc.paper.command.brigadier.argument.ArgumentTypes
import cat.emir.echolib.command.PluginCommand
import cat.emir.echolib.command.getPlayers
import cat.emir.echolib.sendLangMessage
import com.mojang.brigadier.builder.LiteralArgumentBuilder

class SudoCommand(plugin: Echogen) : PluginCommand<Echogen>(plugin) {

    override fun getCommand(): LiteralArgumentBuilder<CommandSourceStack> {
        return command("sudo") {
            requires { it.sender.hasPermission("echogen.sudo") }
            argument("players", ArgumentTypes.players()) {
                argument("message", StringArgumentType.greedyString()) {
                    executes(::sudo)
                }
            }
        }
    }

    fun sudo(ctx: CommandContext<CommandSourceStack>): Int {
        val sender = ctx.source.sender
        val targets = ctx.getPlayers("players") ?: return 1
        val message = StringArgumentType.getString(ctx, "message")

        for (target in targets) {
            target.chat(message)

            sender.sendLangMessage("sudo.executor", listOf(
                "player" to target.name,
                "messsage" to message
            ))
        }

        return 1
    }
}
